test/parallel/test-stream-duplex-end.js

```diff
@@ -0,0 +1,30 @@
+'use strict';
+
+const common = require('../common');
+const assert = require('assert');
+const Duplex = require('stream').Duplex;
+{
+  const stream = new Duplex();
+  assert(stream.allowHalfOpen);
+  stream.on('finish', common.mustNotCall());
+  assert.strictEqual(stream.emit('end'), false);
```

Comment:
I wonder if it's better to use `stream.push(null)` instead of emitting `'end'` directly. We can still verify that there is no listener for `'end'` event with `stream.listenerCount('end')` if wanted.

Reply:
@lpinca Thanks for your review. Would you please tell me why it is better to use `stream.push(null)` instead of `stream.emit('end')`. I tried to use `stream.push(null)` but test didn't emit `'end'` and the function `onend` didn't run.
https://coverage.nodejs.org/coverage-7551de1fc2fbf24e/root/_stream_duplex.js.html#L100

Reply:
That's how the `'end'` event is emitted in a real use case. `readable.push(null)` is used to signal that the stream has ended, you don't emit `'end'` manually.

If you switch the stream in flowing mode with `stream.resume()` and then call `stream.push(null)` it should emit `'end'`. You probably also have to add an implementation for `_read()`.


```js
const stream = new Duplex({ read() {} });
stream.resume();
stream.push(null);
```

Reply:
I fixed it and the codes you indicated emit `'end'`. Thank you for your comments.